[NewOptimizer] Various misc bugfixes

Keno opened the pull request:
Some fixes @vtjnash and I worked on last week that accidentally got dropped from the last roundup commit, as well as one new commit to address a failure I saw on kf/iterate.

Keno committed:
[NewOptimizer] A couple of fixes that got accidentally dropped

Keno committed:
Fix partially specified phinodes of bits types

vtjnash commented on src/codegen.cpp:
```diff
@@ -6610,10 +6611,29 @@ static std::unique_ptr<Module> emit_function(
+#if JL_LLVM_VERSION >= 50000
+                            0,
+#endif
+                            "undef_alloca", /*InsertBefore=*/ctx.ptlsStates);
```
indent

vtjnash commented on src/codegen.cpp:
```diff
@@ -6610,10 +6611,29 @@ static std::unique_ptr<Module> emit_function(
+                            VNUndef =  (llvm::Value*)ConstantPointerNull::get(cast<PointerType>(T_prjlvalue));
+                        } else if (VN->getType()->isPointerTy()) {
+                            PointerType *PTy = cast<PointerType>(VN->getType());
+                            VNUndef = new AllocaInst(PTy->getElementType(),
```
I don't think we do / can correctly set the element type in many cases (especially with union-alloca)

Keno replied:
Ah, fair point, we can just get the size from the union type. Alternatively, we could just emit one alloca big enough for all phis at the top.

Keno commented:
Fixed to use one big undef alloca with appropriate size instead.

vtjnash commented:
Won’t that cause issues for SROA?

Keno commented:
I put a lifetime.end intrinsic at the top of the function, so it's legal for SROA to ignore any loads/stores known to be to that alloca.

Keno committed:
Fix some try/catch issues

Keno committed:
Add missing nospecialize

Keno committed:
Extend undef code to all cases

There's more cases where we need to have the value behave as if it
was an SSA undef. Simply extend the code we have here to all such
cases.

Keno commented:
This has a good number of CI failures. The only one I'm not familiar with is a missing gc root in ccall, which @vtjnash says is fixed by #26732. Will merge once @vtjnash has had a chance to review.
